modules/indexing/src/test/java/org/apache/ignite/internal/processors/cache/index/RejoinWithLostDynamicDdlTest.java

```diff
@@ -0,0 +1,232 @@
+    }
+
+    /** */
+    @Test
+    public void testRejoinWithLostDynamicSchemaInmemoryInactive() throws Exception {
+        testRejoinWithLostDynamicSchema(false, false, false);
+    }
+
+    /** */
+    @Test
+    public void testRejoinWithLostDynamicSchemaPersistentActive() throws Exception {
+        testRejoinWithLostDynamicSchema(true, true, false);
+    }
+
+    /** */
+    @Test
+    public void testRejoinWithLostLostDynamicTablePersistentInactive() throws Exception {
+        testRejoinWithLostDynamicSchema(true, false, false);
+    }
+
+    /** */
+    @Test
+    public void testRejoinWithLostDynamicSchemaPersistentActiveClear() throws Exception {
+        testRejoinWithLostDynamicSchema(true, true, true);
+    }
+
+    /** */
+    @Test
+    public void testRejoinWithLostDynamicSchemaPersistentInactiveClear() throws Exception {
+        testRejoinWithLostDynamicSchema(true, false, true);
+    }
+
+    /**
+     * Tests the scenario when a node rejoins cluster with lost knowladge of previously created dynamic table over
+     * a predefined cache in {@link IgniteConfiguration}.
+     *
+     * @param persistence Flag to test with persistence or in-memory cluster.
+     * @param rejoinActive Flag to rejoin to active or inactive cluster.
+     * @param clearData Flag to clear test node's persistent data before rejoining. Efficient with enabled {@code persistence}.
+     */
+    private void testRejoinWithLostDynamicSchema(
+        boolean persistence,
+        boolean rejoinActive,
+        boolean clearData
+    ) throws Exception {
+        this.persistence = persistence;
+
+        CacheConfiguration<?, ?> cacheCfg = new CacheConfiguration<>("STATIC_CACHE")
+            .setBackups(SERVERS_CNT - 1)
+            .setCacheMode(CacheMode.PARTITIONED)
+            .setAtomicityMode(CacheAtomicityMode.ATOMIC)
+            .setWriteSynchronizationMode(CacheWriteSynchronizationMode.PRIMARY_SYNC);
+
+        staticCaches = new CacheConfiguration<?, ?>[] {cacheCfg};
+
+        startGrids(SERVERS_CNT);
+
+        sqlClient = startClientGrid(G.allGrids().size());
+
+        if (persistence)
+            grid(0).cluster().state(ClusterState.ACTIVE);
+
+        cacheCfg.setName("DYN_CACHE");
+
+        sqlClient.createCache(cacheCfg);
+
+        awaitPartitionMapExchange();
+
+        sql("CREATE TABLE STATIC_TBL(ID INTEGER PRIMARY KEY, VAL VARCHAR) WITH \"CACHE_NAME=STATIC_CACHE\"");
+        sql("CREATE TABLE DYN_TBL(ID INTEGER PRIMARY KEY, VAL VARCHAR) WITH \"CACHE_NAME=DYN_CACHE\"");
+
+        assertEquals(0, sql("SELECT * FROM STATIC_TBL").size());
+        assertEquals(0, sql("SELECT * FROM DYN_TBL").size());
+
+        File persistPath = grid(gridToRestart).context().pdsFolderResolver().fileTree().nodeStorage();
+
+        stopGrid(gridToRestart);
+
+        if (clearData) {
+            if (log.isDebugEnabled())
+                log.debug("Clearing " + persistPath);
+
+            U.delete(persistPath);
```

Comment:
`cleanPersistenceDir(getTestIgniteInstanceName(gridToRestart))`? (but `consistentId` should be configured in `getConfigureation` in this case)